### PR TITLE
Add `cacerts_path` config option

### PR DESCRIPTION
**Pull request by cburman01**
This changes allows the option to specify a custom CA instead of the ones that get compiled into hex. 

Use Case: ssl breaks when using a self signed CA such as in an enterprise environment. 
Current workout: enable unsafe_https. (this is not ideal) 

Reference: https://github.com/hexpm/hex/issues/690
Reference: https://github.com/hexpm/hex/issues/146

This change allows an option to be specified "HEX_ALT_CRT_PATH" which will load that CA instead of the ones that are compiled in. 

I tested as so:
`mix hex.config unsafe_https false  `
`iex -S mix`
`Hex.State.fetch!(:unsafe_https)` Confirm this is false 

Fetch a package from hex:
`Hex.API.Package.get("hexpm", "timex")`

This is the error I get inside a self signed ssl world: 

> {:error,
>  {:failed_connect,
>   [
>     {:to_address, {'hex.pm', 443}},
>     {:inet, [:inet],
>      {:tls_alert,
>       {:unknown_ca,
>        'TLS client: In state certify at ssl_handshake.erl:1666 generated CLIENT ALERT: Fatal - Unknown CA\n'}}}
>   ]}}

Rerun the app with the HEX_ALT_CRT_PATH specified: 
`HEX_ALT_CRT_PATH=/tmp/hl-ca-bundle.crt iex -S mix`
Fetch a package from hex:
`Hex.API.Package.get("hexpm", "timex")`

This is the 200 we get back: 

> {:ok,
>  {200,
>   %{
>     "docs_html_url" => "https://hexdocs.pm/timex/",
>     "downloads" => %{
>       "all" => 10929378,
>       "day" => 63714,
>       "recent" => 3252289,
>       "week" => 306078
>     },
>     "html_url" => "https://hex.pm/packages/timex",
>     "inserted_at" => "2014-04-23T04:00:39Z",
>     "meta" => %{
>       "description" => "Timex is a rich, comprehensive Date/Time library for Elixir projects, with full timezone support via the :tzdata package.\nIf you need to manipulate dates, times, datetimes, timestamps, etc., then Timex is for you!",
>       "licenses" => ["MIT"],
>       "links" => %{
>         "Changelog" => "https://github.com/bitwalker/timex/blob/master/CHANGELOG.md",
>         "GitHub" => "https://github.com/bitwalker/timex"
>       },
>       "maintainers" => []
>     },
>     "name" => "timex",
>     "owners" => [
>       %{
>         "email" => "paulschoenfelder@gmail.com",
>         "url" => "https://hex.pm/api/users/bitwalker",
>         "username" => "bitwalker"
>       }
>     ],
>     "releases" => [
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2019-06-27T20:19:38Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.6.1",
>         "version" => "3.6.1"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2019-06-26T22:41:04Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.6.0",
>         "version" => "3.6.0"
>       },
>       %{ 
>         "has_docs" => true,
>         "inserted_at" => "2019-01-18T16:42:47Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.5.0",
>         "version" => "3.5.0"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-11-03T18:08:04Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.4.2",
>         "version" => "3.4.2"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-09-19T17:17:55Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.4.1",
>         "version" => "3.4.1"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-05-08T23:25:18Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.3.0",
>         "version" => "3.3.0"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-04-14T01:07:14Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.2.2",
>         "version" => "3.2.2"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-02-20T18:39:16Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.2.1",
>         "version" => "3.2.1"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-02-20T18:31:38Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.2.0",
>         "version" => "3.2.0"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2018-01-30T21:36:58Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.25",
>         "version" => "3.1.25"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-07-19T02:54:32Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.24",
>         "version" => "3.1.24"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-07-17T17:07:28Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.23",
>         "version" => "3.1.23"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-07-17T15:01:37Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.22",
>         "version" => "3.1.22"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-07-01T16:38:46Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.21",
>         "version" => "3.1.21"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-06-28T19:33:45Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.20",
>         "version" => "3.1.20"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-06-27T22:17:13Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.19",
>         "version" => "3.1.19"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-06-27T18:08:23Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.18",
>         "version" => "3.1.18"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-06-21T21:21:00Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.17",
>         "version" => "3.1.17"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-06-21T18:14:14Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.16",
>         "version" => "3.1.16"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-05-17T22:22:44Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.15",
>         "version" => "3.1.15"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-02-17T15:22:38Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.13",
>         "version" => "3.1.13"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-02-13T04:43:15Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.12",
>         "version" => "3.1.12"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-02-09T17:58:15Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.11",
>         "version" => "3.1.11"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-02-07T20:16:23Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.10",
>         "version" => "3.1.10"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-02-02T02:12:57Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.9",
>         "version" => "3.1.9"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-01-19T20:57:44Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.8",
>         "version" => "3.1.8"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-01-05T03:32:48Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.7",
>         "version" => "3.1.7"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2017-01-03T21:15:35Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.6",
>         "version" => "3.1.6"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-11-06T19:43:37Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.5",
>         "version" => "3.1.5"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-11-06T05:55:24Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.4",
>         "version" => "3.1.4"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-10-28T15:45:26Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.3",
>         "version" => "3.1.3"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-10-27T17:05:34Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.2",
>         "version" => "3.1.2"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-10-25T21:22:15Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.1",
>         "version" => "3.1.1"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-10-11T21:16:45Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.1.0",
>         "version" => "3.1.0"
>       },
>       %{
>         "has_docs" => true,
>         "inserted_at" => "2016-08-18T14:44:54Z",
>         "url" => "https://hex.pm/api/packages/timex/releases/3.0.8",
>         ...
>       },
>       %{"has_docs" => true, "inserted_at" => "2016-08-15T18:05:46Z", ...},
>       %{"has_docs" => true, ...},
>       %{...},
>       ...
>     ],
>     "repository" => "hexpm",
>     "retirements" => %{
>       "3.1.25" => %{
>         "message" => "Incorrect Elixir version requirement",
>         "reason" => "invalid"
>       }
>     },
>     "updated_at" => "2019-08-11T12:55:00Z",
>     "url" => "https://hex.pm/api/packages/timex"
>   },
>   %{
>     'alt-svc' => 'clear',
>     'cache-control' => 'public, max-age=60',
>     'content-length' => '17024',
>     'content-type' => 'application/vnd.hex+erlang; charset=utf-8',
>     'date' => 'Thu, 21 Nov 2019 13:56:20 GMT',
>     'etag' => 'a17d2c81bc6739e508bcbb6ed2665c79',
>     'last-modified' => 'Sun, 11 Aug 2019 12:55:00 GMT',
>     'server' => 'Cowboy',
>     'strict-transport-security' => 'max-age=31536000',
>     'vary' => 'accept, accept-encoding, accept-encoding',
>     'via' => '1.1 google',
>     'x-ratelimit-limit' => '100',
>     'x-ratelimit-remaining' => '98',
>     'x-ratelimit-reset' => '1574344620',
>     'x-request-id' => 'FdkyDi5fdBNbaLEAPljR'
>   }}}
> 

The case inside Hex.HTTP.SSL.get_ca_certs/0 should be enough to maintain default behaviour if alt_crt_path is not set. 

Any feedback would be appreciated!

### CODE REVIEW

**Comment by cburman01** [comment]
@ericmj Does this seem like a viable solution?

File: lib/hex/state.ex [hunk 1]
```diff
@@ -89,6 +89,11 @@ defmodule Hex.State do
       env: ["HEX_DIFF_COMMAND"],
       config: [:diff_command],
       default: Mix.Tasks.Hex.Package.default_diff_command()
+    },
+    alt_crt_path: %{
```
Comment (ericmj) [inline lib/hex/state.ex]: Please rename this to `cacerts_path`.

File: lib/hex/state.ex [hunk 2]
```diff
@@ -89,6 +89,11 @@ defmodule Hex.State do
...
+      env: ["HEX_ALT_CRT_PATH"],
```
Comment (ericmj) [inline lib/hex/state.ex]: Please rename this to `HEX_CACERTS_PATH`.

**Review by ericmj — COMMENTED** [review]
Looks good, just added two suggestions to change the config and variable names.

File: lib/hex/state.ex (hunk 1 above)
Comment (cburman01) [reply lib/hex/state.ex]: Done

File: lib/hex/state.ex (hunk 2 above)
Comment (cburman01) [reply lib/hex/state.ex]: Done

**Comment by cburman01** [comment]
@ericmj Thanks! Much better var names :)

File: lib/hex/state.ex [hunk 3]
```diff
@@ -89,6 +89,11 @@ defmodule Hex.State do
       env: ["HEX_DIFF_COMMAND"],
       config: [:diff_command],
       default: Mix.Tasks.Hex.Package.default_diff_command()
+    },
+    cacerts_path: %{
+      env: ["HEX_CACERTS_PATH"],
+      default: nil,
+      config: [:alt_crt_path]
```
Comment (wojtekmach) [inline lib/hex/state.ex]: should it be this?

```suggestion
      config: [:cacerts_path]
```

File: lib/hex/state.ex [hunk 4]
```diff
@@ -89,6 +89,11 @@ defmodule Hex.State do
       env: ["HEX_DIFF_COMMAND"],
       config: [:diff_command],
       default: Mix.Tasks.Hex.Package.default_diff_command()
+    },
+    cacerts_path: %{
```
Comment (wojtekmach) [inline lib/hex/state.ex]: Could you also document this new config key? https://github.com/hexpm/hex/blob/master/lib/mix/tasks/hex.config.ex#L19

File: lib/hex/state.ex (hunk 3 above)
Comment (cburman01) [reply lib/hex/state.ex]: Good Catch! Done.

File: lib/hex/state.ex (hunk 4 above)
Comment (cburman01) [reply lib/hex/state.ex]: Great! This is done: 50ec5ec56a8200e103173eb03e754384efc92552

**Review by wojtekmach — APPROVED** [review]

File: lib/hex/state.ex (hunk 4 above)
Comment (wojtekmach) [reply lib/hex/state.ex]: I ended up changing docs slightly, what do you think?

Comment (cburman01) [reply lib/hex/state.ex]: Yeah, much clearer. Thanks!

**Review by ericmj — APPROVED** [review]
Beautiful! Thanks for the contribution.

**Comment by nwoodr94** [comment]
Can someone explain how this works and/or please add it to the Getting Started docs? I know how to configure certs for pip, npm, etc. but I am really struggling here.

I've tried adding `cacerts_path: "../certs/mycert.pem"` to my `./config/config.exs` under tailwind and esbuild, passing in a bunch of different options to my `mix phx.server` command, but nothing is working.

**Comment by ericmj** [comment]
1. and 2. is not how you configure this, 3. should be the way to do it. Can you show the commands you are running? Have you tried setting `HEX_CACERTS_PATH`?

**Comment by nwoodr94** [comment]
> 1. and 2. is not how you configure this, 3. should be the way to do it. Can you show the commands you are running? Have you tried setting `HEX_CACERTS_PATH`?

`HEX_CACERTS_PATH` is an invalid key: 
```
> mix hex.config HEX_CACERTS_PATH C:\Users\user\path\to\my\certs\pem.pem
> ** (Mix) Invalid key HEX_CACERTS_PATH
```

The docs say `Can be overridden by setting the environment variable HEX_CACERTS_PATH`. What does that mean?

I'm able to set `cacerts_path`:
```
> mix hex.config cacerts_path
>"C:\Users\user\path\to\my\certs\pem.pem" (using `c:/Users/user/.hex/hex.config`)
```

I checked the `hex.config` noted in paranthesis:
```
{cacerts_path,<<"C:\Users\user\path\to\my\certs\pem.pem">>}.
```

**Comment by wojtekmach** [comment]
> How do I set HEX_CACERTS_PATH?

If you are using cmd.exe, it would be:

```
> set HEX_CACERTS_PATH=C:\Users\user\path\to\my\certs\pem.pem
> mix phx.server
```

You can learn more about environment variables here: https://en.wikipedia.org/wiki/Environment_variable

> The docs say Can be overridden by setting the environment variable HEX_CACERTS_PATH. What does that mean?

Hex looks up configuration values using precedence explained here: https://hexdocs.pm/hex/Mix.Tasks.Hex.Config.html#module-config-overrides

Does that help?

**Comment by nwoodr94** [comment]
Thanks for clarifying. I'm brand new to Phoenix so I didn't suspect I needed a Powershell environment variable. 

I've tried this using both a pem and cert.
`$env:HEX_CACERTS_PATH = 'C:\Users\user\path\to\my\certs\pem.pem'`
`$env:HEX_CACERTS_PATH = 'C:\Users\user\path\to\my\certs\crt.crt'`

I'm still getting the same TLS error
`[notice] TLS :client: In state :certify at ssl_handshake.erl:2134 generated CLIENT ALERT: Fatal - Unknown CA`

Now each of my System Configuration, Project Configuration, and Global Configuration all point to the same absolute path to my `.pem` file.

I'm working with a fresh install of Elixer and the Erlang VM - do I need some configuration at this level?

**Comment by nwoodr94** [comment]
Following up on this, we had some kind of firewall rules on `github.com` and `registry.npmjs.org`, at least for `mix`.

Why our firewall blocked traffic for `mix` when we otherwise hit github and other registries everyday through `pip`, `npm` and `apt` is beyond me, but after an admin temporarily disabled my VPN connection, Phoenix started right up as expected.

**Comment by James-E-A** [comment]
I'm also seeing this issue still.

Node respects `NODE_EXTRA_CA_CERTS`, and Python mostly respects `REQUESTS_CA_BUNDLE`, but `HEX_CACERTS_PATH` set to the exact same value is apparently being ignored by all Elixir tooling, particularly *mix deps\.get*...

**Comment by Friph** [comment]
I'm also have still the same issue on macOS with nixOS multi installer and tailwind install.
In my flake.nix I export HEX_CACERTS_PATH
image

When I will install tailwind with mix tailwind.install I get the following error:

image
My hex ca_certs is correctly set:
image

Tailwind calls for install fetch_body with :public_key.cacerts_get():

image

I think the CA should be ok, because HTTPPoison (Hackney) works with it.
Is there still a solution now?

**Comment by maennchen** [comment]
@Friph The environment variable only works for Hex itself. `mix tailwind.install` is part of the `tailwind` package and does not read that environment variable.

You'll have to either add an option to the `tailwind` package or reconfigure Erlang CACerts.

You can reconfigure Erlang CACerts with adding an option `erl -public_key cacerts_path '"/path/to/certs.pem"'`. For Mix, you can set this through the env variable `ELIXIR_ERL_OPTIONS="-public_key cacerts_path 'path/to/certs.pem'"`, which is passed through to `erl`.

**Comment by Friph** [comment]
Thanks for your help @maennchen - We use actually OTP 25. So the erl opts doesnt work.
But now I have a workaround til we upgrade to OTP > 27.2 : https://github.com/phoenixframework/tailwind/issues/138#issuecomment-4509800476

**Pull request merged**